### PR TITLE
Add support for aggregating exposure keys

### PR DESCRIPTION
**Pull request by wescarr**

### CODE REVIEW

File: aggregate/aggregate.go [hunk 1]
```diff
@@ -361,3 +394,44 @@ func Tokens(ctx context.Context, c *config.Config, s *storage.Client, throttle i
...
+
+// ExposureKeys handles aggregating all the input expsorue keys in a holding bucket and publishing to
+// a publish bucket
+func ExposureKeys(ctx context.Context, c *config.Config, s *storage.Client, throttle int64) error {
+	readers, objects, err := getObjectReaders(ctx, s.Bucket(c.ExposureKeysHoldingBucket))
+	if err != nil {
+		return err
+	}
+
+	records, err := getRecords(readers, true, 3)
```
Comment (joshgummersall) [inline aggregate/aggregate.go:406]: The `3` here is the expected number of columns in the CSV. Looks like it should be `2` based on the struct above.

**Review by joshgummersall — COMMENTED** [review]
Looks good to me, save for one small comment. What's the bug you're hitting?

File: aggregate/aggregate.go (hunk 1 above)
Comment (wescarr) [reply aggregate/aggregate.go:406]: Good catch

File: aggregate/aggregate.go [hunk 2]
```diff
@@ -361,3 +394,44 @@ func Tokens(ctx context.Context, c *config.Config, s *storage.Client, throttle i
...
+	keys := recordsToExposureKeys(records)
+
+	// Control goroutine concurrency
+	sem := make(chan bool, throttle)
+	sem <- true
+
+	if len(keys) != 0 {
+		name := fmt.Sprintf("%v.csv", time.Now().Unix())
+		o := s.Bucket(c.ExposureKeysPublishedBucket).Object(name)
+		err := writeRecords(ctx, o, expsoreKeysToRecords(c, keys))
```
Comment (wescarr) [inline aggregate/aggregate.go:420]: Does this need to be wrapped in Go routine like the other functions?

File: aggregate/aggregate.go (hunk 1 above)
Comment (wescarr) [reply aggregate/aggregate.go:406]: Actually they have `3`, `key_data,rolling_start_number,verified`

File: aggregate/aggregate_test.go [hunk 3]
```diff
@@ -51,6 +53,18 @@ func TestTokens(t *testing.T) {
 	}
 }
 
+func TestExposureKeys(t *testing.T) {
```
Comment (wescarr) [inline aggregate/aggregate_test.go:56]: FYI, when I run this locally, it doesn't seem to process files in the holding bucket, not sure why

File: aggregate/aggregate.go [hunk 4]
```diff
@@ -154,6 +159,23 @@ func recordsToTokens(rs records) []token {
 	return tokens
 }
 
+func recordsToExposureKeys(rs records) []exposureKey {
+	keys := []exposureKey{}
+
+	for _, r := range rs {
+		if rollingStartNumber, err := strconv.ParseInt(r[1], 10, 64); err == nil {
```
Comment (joshgummersall) [inline aggregate/aggregate.go:166]: This err isn’t logged and could be causing problems with the test.

File: aggregate/aggregate.go (hunk 2 above)
Comment (joshgummersall) [reply aggregate/aggregate.go:420]: I think the other methods that write data use one go routine for each file being produced. This only produces a single file though right?

File: aggregate/aggregate.go [hunk 5]
```diff
@@ -361,3 +394,44 @@ func Tokens(ctx context.Context, c *config.Config, s *storage.Client, throttle i
...
+	records, err := getRecords(readers, true, 3)
+	if err != nil {
+		log.Println(err)
+		return err
+	}
+
+	keys := recordsToExposureKeys(records)
+
+	// Control goroutine concurrency
+	sem := make(chan bool, throttle)
```
Comment (joshgummersall) [inline aggregate/aggregate.go:415]: You can remove this semaphore stuff as you aren’t spinning up any go routines.

Comment (joshgummersall) [reply aggregate/aggregate.go:415]: Actually it’s an argument to the archive method. Remove the next line though where you write true into it - you aren’t pulling that value out in this method so it’s unnecessary.

File: aggregate/aggregate.go (hunk 2 above)
Comment (wescarr) [reply aggregate/aggregate.go:420]: Correct

File: aggregate/aggregate.go (hunk 4 above)
Comment (wescarr) [reply aggregate/aggregate.go:166]: I'm using the same pattern as the other parsing methods, note the check for `err == nil`

Comment (joshgummersall) [reply aggregate/aggregate.go:166]: Right - just saying maybe for testing there's an error here that's hidden.

File: aggregate/aggregate.go (hunk 2 above)
Comment (wescarr) [reply aggregate/aggregate.go:420]: Correct

**Comment by joshgummersall** [comment]
Any luck figuring out the testing thing?

**Comment by wescarr** [comment]
@joshgummersall seems fine now, test runs properly. I must have run before I finalized everything

**Review by joshgummersall — APPROVED** [review]
Lgtm

**Pull request merged**